[#38672] Project members shown repetitively in member page if a group filter is active

fraber opened the pull request:
https://community.openproject.org/work_packages/38672

… (and the user is in several groups that have a membership role).

Added a "select distinct * from (...)" around the results of MemberQuery in order to eliminate multiple rows returned by the group filter in MemberQuery if a user belongs to multiple groups.
Also added a brand new spec for MemberQuery testing the happy path and the regression of this bug.

fraber committed:
[#38672] Project members are shown repetitively in member page if a group filter is active (and the user is in several groups that have a membership role)

https://community.openproject.org/work_packages/38672
Added a "select distinct * from (...)" around the results
of MemberQuery in order to eliminate multiple rows returned
by the group filter in MemberQuery if a user belongs to
multiple groups.
Also added a brand new spec for MemberQuery testing the
happy path and the regression of this bug.

fraber committed:
Fixed ordering issues after applying a DISTINCT to query results.
Moved DISTINCT into the inner part of the "select * from (...)"
to be compatible with Relations adding inner joins to the base
query.

wielinde commented on spec/models/queries/members/member_query_spec.rb:
```diff
@@ -0,0 +1,91 @@
+  context 'with a user but without filters' do
+    before do
+      # We need to write the membership to the DB to count results
+      member.save
```
`.save` should not be necessary

wielinde commented on spec/models/queries/members/member_query_spec.rb:
```diff
@@ -0,0 +1,91 @@
+    describe '#results' do
+      # Check that the query SQL is the same as the manual SQL
+      it 'is the same as getting all the members' do
+        expect(instance.results.to_sql).to eql base_scope.to_sql
```
hmm, you are comparing one intransparent magic with another intransparent magic. The test is not obvious. I am not a big fan of comparing SQL here. Can't we simply check what the results are and in which order they appear, etc.?

wielinde commented on spec/models/queries/members/member_query_spec.rb:
```diff
@@ -0,0 +1,91 @@
+  let(:admin) { create(:admin) }
+  let(:instance) { described_class.new(user: admin) }
+  # This is the MemberQuery.default_scope with an admin user
+  let(:base_scope) { Member.from(Member.all.distinct, :members).order(id: :desc) }
```
Can we find a better name for `base_scope`? Or drop the idea of comparing SQL all together?

wielinde commented on spec/models/queries/members/member_query_spec.rb:
```diff
@@ -0,0 +1,91 @@
+
+    describe '#results' do
+      it 'returns the single test user if no filters are specified' do
+        expect(instance.results.count).to eq 1
```
maybe adding another `expect` to check that the expected user is part of the result set?

wielinde commented on spec/models/queries/members/member_query_spec.rb:
```diff
@@ -0,0 +1,91 @@
+    let(:instance) { described_class.new(user: admin).where(:group, '=', group1.id) }
+
+    before do
+      member.save
```
no need for `.save`

wielinde commented on spec/models/queries/members/member_query_spec.rb:
```diff
@@ -0,0 +1,91 @@
+    before do
+      member.save
+      # Add the admin user to the two groups
+      Groups::UpdateService.new(user: admin, model: group1).call(user_ids: Array(admin.id))
```
```suggestion
      Groups::UpdateService.new(user: admin, model: group1).call(user_ids: [admin.id])
```

wielinde commented on spec/models/queries/members/member_query_spec.rb:
```diff
@@ -0,0 +1,91 @@
+      member.save
+      # Add the admin user to the two groups
+      Groups::UpdateService.new(user: admin, model: group1).call(user_ids: Array(admin.id))
+      Groups::UpdateService.new(user: admin, model: group2).call(user_ids: Array(admin.id))
```
same as above

wielinde reviewed (changes requested):
@fraber Please find may comments in the code. You propably also want Jens to review this PR.

fraber committed:
Fixed Wielands comments in pull request #38672

fraber commented:
Hi @ulferts , Wieland mentioned it would be a good idea if you could have a look at this PR.

ulferts commented:
Hi @fraber applying `DISTINCT` to a query has the disadvantage in PostgreSQL that every field listed in the `ORDER BY` of the query will have to appear in the list that is checked DISTINCT on (e.g. `SELECT DISTINCT id FROM members ORDER BY created_at` will break). Applying a distinct will in my experience result in added complexity when one tries to either remove the distinct again to get around the limitation or one tries to manipulate to projection part.

In this case, the fix might be simpler though I don't know if anything will break:  
In my very brief test, removing all of https://github.com/opf/openproject/blob/dev/app/models/queries/members/filters/group_filter.rb#L32-L39 fixed the problem. I assume (don't know any more since it has been a while) that the included `Queries::Filters::Shared::GroupFilter` used to be implemented differently. The uselessly overwritten `joins` method strongly suggests that.

jngrb commented:
@wielinde, @fraber, @ulferts, as creator of the bug report [OPENPROJECT 38672](https://community.openproject.org/projects/openproject/work_packages/38672), I would like to know if this PR is ready to be merged or if there are still changes to be made.

wielinde commented:
@jngrb No, unfortunately it is not ready to merge. The review feedback from @ulferts is valid and needs to be addressed first by @fraber.

fraber commented:
@jngrb , sorry, vacation time has started. I'd expect me/us to finish the issue in about 3-4 weeks (mid-August).

fraber committed:
Changed the scope for Queries::Members::Filters::GroupFilter
based on comment from @ulferts, but including a .joins(:principals)
because of a breaking API when calling memberships with a group
filter. Both the breaking API case and the regression case are
covered by spec/models/queries/members/member_query_spec.rb

fraber commented:
@ulferts , I implemented [your proposed change](https://github.com/opf/openproject/pull/10599#issuecomment-1119760161) for Queries::Members::Filters::GroupFilters above. Your suggestions successfully fixed the reported issue in the /projects/ID/members page.

Then I did quite a bit of manual testing in all places where GroupFilter was used and found that your proposed change broke the API call /api/v3/memberships when using with filters=[{"group":{"operator":"=","values":["17"]}}]:
```
PG::UndefinedTable: ERROR:  missing FROM-clause entry for table "users"
LINE 1: SELECT COUNT(*) FROM "members" WHERE (users.id IN (SELECT "u...
```

So I've added the required **.joins(:principal)** back to Queries::Members::Filters::GroupFilter,
together with the other options from the base class to handle from, joins and left_outer_joins (not sure about this...)

I did another round of testing to make sure that both the breaking API and the regression case are covered by specs (spec/models/queries/members/member_query_spec.rb).

After a rebase to current dev I got an issue manually testing in /projects/<id>/members with the same error message as in the API call above. The .joins(:principal) fixed this in the same way as it did with the API call. However, this issue disappeared after a while, so I couldn't write a spec for it. I understand the issue would be detected by the spec/feature/members/ suite.
